jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -845,6 +877,13 @@ Object runNativeMethod(Addressable downcallAddr, SegmentAllocator segmtAllocator
 		long returnVal;
 		/*[IF JAVA_SPEC_VERSION >= 22]*/
 		HeapArgInfo info = heapArgInfo.get();
+		/* Notify to reset the heap argument information for the current thread prior to downcall
+		 * so as to update the base object of different heap arguments passed by the same handler
+		 * within the current thread the next time when performing the downcall.
+		 */
+		if (info != null) {
+			info.index = 0;
+		}
```

Comment:
I think we want to define this method in `HeapArgInfo`:
```java
		void clear() {
			Arrays.fill(bases, null);
			index = 0;
		}
```
to be used instead of just resetting `index`. Here, that call should occur in a `finally` block surrounding the call to `invokeNative()`. By clearing the array of references, we avoid possible memory leaks (we don't want a strong reference held in `HeapArgInfo.bases` in the thread-local).

Reply:
It makes sure in this way but I am not sure whether it still works with the finally block in the case of exception triggered by upcall (in which the exception is directly thrown out in native and might never get back to the java code). I will need to verify the changes to see what happens.

Reply:
I finished verifying with the FFI test suites (OpenJ9 & Jtreg) which works good as expected.